File: leetcodepractice/delete-node-in-a-bst/delete-node-in-a-bst.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} key
 * @return {TreeNode}
 */
var deleteNode = function(root, key) {
    if (!root) return null;

    if (key < root.val){
        root.left = deleteNode(root.left, key);
    } else if (root.val < key){
        root.right = deleteNode(root.right, key);
    } else {
        if (!root.left) {
            return root.right
        } else if (!root.right){
            return root.left
        } else {
            const minNode = getMinNode(root);
            root.val = minNode.val;
            root.right = deleteNode(root.right, minNode.val);
        }
    }

    return root;
};

const getMinNode = (root) => {
    let curr = root.right
    while (curr && curr.left){
        curr = curr.left
    }
    return curr
}

File: leetcodepractice/450-delete-node-in-a-bst/delete-node-in-a-bst.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @param {Integer} key
# @return {TreeNode}

def get_min_node(root)
    curr = root

    while curr && curr.left do
        curr = curr.left
    end

    return curr
end

def delete_node(root, key)
    return nil if !root

    if root.val < key
        root.right = delete_node(root.right, key)
    elsif root.val > key
        root.left = delete_node(root.left, key)
    else
        if !root.left
            return root.right
        elsif !root.right
            return root.left
        else
            min_node = get_min_node(root.right)
            root.val = min_node.val
            root.right = delete_node(root.right, min_node.val)
        end
    end

    return root
end
